security/src/main/scala/pl/touk/nussknacker/ui/security/oauth2/CachingOAuth2Service.scala

```diff
@@ -1,37 +1,53 @@
+import scala.util.{Failure, Success}
 
 class CachingOAuth2Service[
   UserInfoData,
   AuthorizationData <: OAuth2AuthorizationData
 ](delegate: OAuth2Service[UserInfoData, AuthorizationData],
-  configuration: OAuth2Configuration)
+  configuration: OAuth2Configuration,
+  ticker: Ticker = Ticker.systemTicker())
  (implicit ec: ExecutionContext) extends OAuth2Service[UserInfoData, AuthorizationData] with LazyLogging {
 
-  protected val authorizationsCache = new DefaultAsyncCache[String, (UserInfoData, Deadline)](CacheConfig(new ExpiryConfig[String, (UserInfoData, Deadline)]() {
-    override def expireAfterWriteFn(key: String, value: (UserInfoData, Deadline), now: Deadline): Option[Deadline] = Some(value._2)
-  }))
+  protected val authorizationsCache = new DefaultCache[String, (UserInfoData, Instant)](CacheConfig(new ExpiryConfig[String, (UserInfoData, Instant)]() {
+    override def expireAfterWriteFn(key: String, value: (UserInfoData, Instant), now: Deadline): Option[Deadline] =
+      Some(Deadline.now + FiniteDuration(Duration.between(Instant.now(), value._2).toNanos, TimeUnit.NANOSECONDS))
+  }), ticker = ticker)
 
   def obtainAuthorizationAndUserInfo(authorizationCode: String, redirectUri: String): Future[(AuthorizationData, UserInfoData)] = {
     delegate.obtainAuthorizationAndUserInfo(authorizationCode, redirectUri).map { case (authorization, userInfo) =>
       authorizationsCache.put(authorization.accessToken) {
-        Future((userInfo, Deadline.now + authorization.expirationPeriod.getOrElse(defaultExpiration)))
+        val expiration = authorization.expirationPeriod.getOrElse(defaultExpiration)
+        (userInfo, Instant.now() plusNanos expiration.toNanos)
       }
       (authorization, userInfo)
     }
   }
 
-  def checkAuthorizationAndObtainUserinfo(accessToken: String): Future[(UserInfoData, Option[Deadline])] = {
-    authorizationsCache.getOrCreate(accessToken) {
-      delegate.checkAuthorizationAndObtainUserinfo(accessToken).map {
-        case (userInfo, expiration) => (userInfo, expiration.getOrElse(Deadline.now + defaultExpiration))
-      }
-    }.map { case (userInfo, expiration) => (userInfo, Some(expiration)) }
+  def checkAuthorizationAndObtainUserinfo(accessToken: String): Future[(UserInfoData, Option[Instant])] = {
+    val userInfo = authorizationsCache.get(accessToken) match {
+      case Some(value) =>
+        Future.successful(value)
+      case None =>
+        val f = delegate.checkAuthorizationAndObtainUserinfo(accessToken).map {
+          case (userInfo, expiration) => (userInfo, expiration.getOrElse(Instant.now() plusNanos defaultExpiration.toNanos))
+        }
+        f.onComplete {
+          case Success(value) => authorizationsCache.put(accessToken)(value)
+          case _ => ()
+        }
+        f
+    }
+    userInfo.map { case (userInfo, expiration) => (userInfo, Some(expiration)) }
   }
 
   private lazy val defaultExpiration = configuration.defaultTokenExpirationTime
```

Comment:
Shit. It looks like we're messing expiration time and expiration period / duration. In this place by expiration time, we think about duration and in JWT object we call that way instant representation

Reply:
I'm all for it.